odoo/addons/test_read_group/tests/test_related.py

```diff
@@ -0,0 +1,107 @@
+
+
+class TestRelated(common.TransactionCase):
+    ''' Test grouping by related non-stored fields. '''
+
+    def test_related(self):
+        Partner = self.env['res.partner']
+        Model1 = self.env['test_read_group.related1']
+        Model2 = self.env['test_read_group.related2']
+        Model3Access00 = self.env['test_read_group.related3.access00']
+        Model3Access01 = self.env['test_read_group.related3.access01']
+        Model3Access10 = self.env['test_read_group.related3.access10']
+        Model3Access11 = self.env['test_read_group.related3.access11']
+        Model3All = (Model3Access00, Model3Access01, Model3Access10, Model3Access11)
+
+        country = self.env.ref('base.ru')
+        state1 = self.env['res.country.state'].create({
+            'name': 'Bashkortostan',
+            'code': 'RU-BA',
+            'country_id': country.id,
+        })
+        state2 = self.env['res.country.state'].create({
+            'name': 'Tatarstan',
+            'code': 'RU-TA',
+            'country_id': country.id,
+        })
+
+        for model in Model3All:
+            for state in (state1, state2, None):
+                for name in ('Ivanov', 'Petrov'):
+                    model.create({
+                        'partner_id': Partner.create({
+                            'name': 'Ivan %s %s' % (name, state and 'from ' + state.name or 'who forgot his origins'),
+                            'country_id': country.id,
+                            'state_id': state and state.id,
+                        }).id
+                    })
+            model.flush()
+
+        # sudo mode
+        for model in Model3All:
+            self.make_test(model, 'state_code')
+            self.make_test(model, 'state_code2')
+            self.make_test(model, 'state_stored_code')
+
+        for model in (Model2,) + Model3All:
+            self.make_test(model, 'country_code2')
+            self.make_test(model, 'partner_city')
+
+        for model in (Model1, Model2) + Model3All:
+            self.make_test(model, 'country_code')
+
+        # Demo user has access to Model3*, but not to Model2, Model1
+        demo = self.env.ref('base.user_demo')
+        self.assertTrue(demo.has_group('base.group_user'), 'Demo user is supposed to have Employee access rights')
+        self.assertFalse(demo.has_group('base.group_system'), 'Demo user is not supposed to have Administrator access rights')
+        # compute_sudo=False, auto_join=True
+        model = Model3Access01.with_user(demo)
+        # To read state_code one must pass via related1_id which has auto_join=False
+        with self.assertRaises(AccessError):
+            self.make_test(model, 'state_code')
+        # to read the code one must pass via:
+        # * related2_id (auto_join=True) -> OK
+        # * test_read_group.related2::state_id (compute_sudo=True) -> we get sudo, so all next checks are not needed
+        self.make_test(model, 'state_code2')
```

Comment:
The two fields ``state_code`` and ``state_code2`` look the same and are essentially related to  ``partner_id.state_id.code``, but ``state_code2`` is passed via intermediate field ``state_id`` in model ``test_read_group.related2`` and implicitly gets compute_sudo

```
class Related2(models.Model):
    _name = 'test_read_group.related2'
    _inherits = {'test_read_group.related1': 'related1_id'}
    _description = 'related2'

    related1_id = fields.Many2one('test_read_group.related1', required=True)
    state_id = fields.Many2one('res.country.state', related='partner_id.state_id')
```

Reply:
I updated first implementation and ``state_code`` doesn't raise access error, because it has another field with auto_join=True